cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -45,67 +49,105 @@ constexpr static inline const float kAdjustCentersWeight = 7.0f;
  * NB: no minibatch splitting is done here, it may require large amount of temporary memory (n_rows
  * * n_cluster * sizeof(float)).
  *
+ * @tparam LabelT label type
+ *
  * @param handle
  * @param[in] centers a pointer to the row-major matrix of cluster centers [n_clusters, dim]
  * @param n_clusters number of clusters/centers
  * @param dim dimensionality of the data
  * @param[in] dataset a pointer to the data [n_rows, dim]
+ * @param[in] dataset_norm pointer to the precomputed norm (for L2 metrics only) [n_rows]
  * @param n_rows number samples in the `dataset`
  * @param[out] labels output predictions [n_rows]
  * @param metric
  * @param stream
  * @param mr (optional) memory resource to use for temporary allocations
  */
+template <typename IdxT, typename LabelT>
 inline void predict_float_core(const handle_t& handle,
                                const float* centers,
                                uint32_t n_clusters,
                                uint32_t dim,
                                const float* dataset,
+                               const float* dataset_norm,
                                size_t n_rows,
```

Comment:
Should `n_rows` and all variables/arrays referring to the dataset indices/sizes be the same type as indices, i.e. `IdxT`?

Reply:
For now, I left `dim` and `n_clusters` in `uint32_t`, replaced `uint64_t` with `IdxT`, and solved a few overflows (I've tested with very large datasets to debug overflows).